src/Avalonia.Native/NotificationManager.cs

```diff
@@ -0,0 +1,67 @@
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using Avalonia.Controls.Notifications;
+using Avalonia.Native.Interop;
+
+namespace Avalonia.Native
+{
+    public class AvaloniaNativeNotificationManager : INotificationManager
+    {
+        private delegate void NotificationCloseDelegate(int id);
+        private delegate void NotificationActionDelegate(int id);
+    
+        private readonly IAvnNotificationManager _native;
+        private readonly Dictionary<int, INotification> _notifications;
+        // ReSharper disable once PrivateFieldCanBeConvertedToLocalVariable
+        private readonly NotificationCloseDelegate _notificationCloseCallback;
+        // ReSharper disable once PrivateFieldCanBeConvertedToLocalVariable
+        private readonly NotificationActionDelegate _notificationActionCallback;
+        private readonly Random _idGenerator;
+ 
+        public AvaloniaNativeNotificationManager(IAvnNotificationManager native)
+        {
+            _native = native;
+            _notifications = new Dictionary<int, INotification>();
+            _notificationCloseCallback = NotificationOnClose;
+            _notificationActionCallback = NotificationOnAction;
+            _native.CloseCallback = Marshal.GetFunctionPointerForDelegate(_notificationCloseCallback);
```

Comment:
Don't use function pointers with Avalonia.Native. Create a COM interface and implement it on the C# side instead. We are automatically generating native-to-managed call proxies for interfaces with `Events` suffix.

The preferred way would be to *not* use an `int` identifier field in AvnNotification but introduce
```c
AVNCOM(IAvnNotificationEvents, 21) : IUnknown
{
    virtual void OnAction() = 0;
    virtual void OnClose() = 0;
};
```
and create a new instance of a class implementing that interface for each `ShowNotification` call and pass it along with other notification info, then keep it in `ComPtr<IAvnNotificationEvents>` on the native side. That way the callback life time would be managed automatically and we won't need to keep the GC-rooted dictionary with all of the notifications.

Reply:
> The preferred way would be to not use an int identifier field in AvnNotification

I guess once we implement replacing of notifications it still makes sense to have some id field to identify the OSX notification we want to replace.

> then keep it in ComPtr<IAvnNotificationEvents> on the native side

That means I have to store a vector of all the `ComPtr<IAvnNotificationEvents>` and release them when the notification closes? AFAIU you want the interface of `IAvnNotificationManager` to be `ShowNotification(AvnNotification* notification, IAvnNotificationCallbacks* callbacks)`, correct?

Reply:
+

Reply:
The user can show multiple notifications. I need to store their corresponding callbacks somewhere in order to call them from the global `NotificationCenterDelegate`. I could plug them into the `NSUserNotification.userInfo` somehow, I suppose. `userInfo` expects a map of `id`, though (which is essentially a pointer to `objc_object`). So I guess I need to find some ugly workaround of putting a `ComPtr` in there...

Reply:
>userInfo expects a map of id

it seems that OSX expects the contents of userInfo to be serializable. I'm afraid that might need to change the API design to accomodate OSX notifications properly.

Reply:
What API design are you thinking of? For now, we could also keep it with the `Dictionary` on C# and the function pointers. And revisit it in the future once we figured out a good/better way of designing this API.

Reply:
> For now, we could also keep it with the Dictionary on C# and the function pointers.

That can and will lead to memory leaks.